instrumentation/dubbo/src/main/java/brave/dubbo/TracingFilter.java

```diff
@@ -90,73 +109,86 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
     Kind kind = rpcContext.isProviderSide() ? Kind.SERVER : Kind.CLIENT;
     final Span span;
+    DubboClientRequest clientRequest = null;
+    DubboServerRequest serverRequest = null;
     if (kind.equals(Kind.CLIENT)) {
-      // When A service invoke B service, then B service then invoke C service, the parentId of the
-      // C service span is A when read from invocation.getAttachments(). This is because
-      // AbstractInvoker adds attachments via RpcContext.getContext(), not the invocation.
-      // See org.apache.dubbo.rpc.protocol.AbstractInvoker(line 141) from v2.7.3
-      Map<String, String> attachments = RpcContext.getContext().getAttachments();
-      DubboClientRequest request = new DubboClientRequest(invocation, attachments);
-      span = tracer.nextSpan(clientSampler, request);
-      injector.inject(span.context(), request);
+      clientRequest = new DubboClientRequest(invoker, invocation);
+      span = clientHandler.handleSend(clientRequest);
     } else {
-      DubboServerRequest request = new DubboServerRequest(invocation, invocation.getAttachments());
-      TraceContextOrSamplingFlags extracted = extractor.extract(request);
-      span = nextSpan(extracted, request);
-    }
-
-    if (!span.isNoop()) {
-      span.kind(kind);
-      String service = DubboParser.service(invocation);
-      String method = DubboParser.method(invocation);
-      span.name(service + "/" + method);
-      DubboParser.parseRemoteIpAndPort(span);
-      span.start();
+      serverRequest = new DubboServerRequest(invoker, invocation);
+      span = serverHandler.handleReceive(serverRequest);
     }
 
     boolean deferFinish = false;
     Scope scope = currentTraceContext.newScope(span.context());
+    Result result = null;
     Throwable error = null;
     try {
-      Result result = invoker.invoke(invocation);
+      result = invoker.invoke(invocation);
       error = result.getException();
       Future<Object> future = rpcContext.getFuture(); // the case on async client invocation
       if (future instanceof CompletableFuture) {
         deferFinish = true;
-        ((CompletableFuture<?>) future).whenComplete((v, t) -> {
-          if (t != null) onError(t, span);
-          span.finish();
-        });
+        if (clientRequest != null) {
+          ((CompletableFuture<?>) future).whenComplete(
+            new FinishClientRequest(clientHandler, clientRequest, result, span));
+        } else {
+          ((CompletableFuture<?>) future).whenComplete(
+            new FinishServerRequest(serverHandler, serverRequest, result, span));
+        }
       }
       return result;
     } catch (Throwable e) {
       propagateIfFatal(e);
       error = e;
       throw e;
     } finally {
-      if (error != null) onError(error, span);
-      if (!deferFinish) span.finish();
+      if (!deferFinish) {
+        if (clientRequest != null) {
+          clientHandler.handleReceive(new DubboClientResponse(clientRequest, result, error), span);
```

Comment:
this tension is ok as it is somewhat uncommon for client and server to use the same handler/filter and types.